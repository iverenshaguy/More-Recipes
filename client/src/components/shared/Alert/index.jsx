import NormalAlert from './NormalAlert';

export default { NormalAlert };
