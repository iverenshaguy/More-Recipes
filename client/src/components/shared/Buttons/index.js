import AddRecipeBtn from './AddRecipeBtn';

export default { AddRecipeBtn };
