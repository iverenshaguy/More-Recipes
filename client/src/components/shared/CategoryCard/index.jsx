import React from 'react';
import { Link } from 'react-router-dom';
import { Card, CardImg, CardImgOverlay } from 'reactstrap';
import './CategoryCard.scss';

const CategoryCard = () => (
  <div className="col-xs-12 col-sm-6 col-md-4 col-lg-3 mt-4 category-card">
    <Link to="/">
      <Card className="m-0 p-0 card-image">
        <CardImg
          bottom
          width="100%"
          className="img-fluid"
          src="images/jollof-rice-img.jpg"
          alt="recipe"
        />
        <CardImgOverlay className="text-center card-image-overlay text-white px-2 recipe-rating">
          <h5 className="mb-1 pt-3">Category</h5>
        </CardImgOverlay>
      </Card>
    </Link>
  </div>
);

export default CategoryCard;
