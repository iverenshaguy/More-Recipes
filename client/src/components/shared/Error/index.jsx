import React from 'react';
import PropTypes from 'prop-types';
import './Error.scss';

/**
 * @exports
 * @function Error
 * @param {object} props
 * @returns {JSX} Error
 */
const Error = (props) => {
  const { type } = props;
  return (
    <div className="container align-items-center justify-content-center" id="error-wrapper">
      <div className="container">
        <h3 className="text-center">{type === 404 ? 'Page Not Found' : 'Something happened, please try again'}</h3>
      </div>
    </div>
  );
};

Error.propTypes = {
  type: PropTypes.number
};

Error.defaultProps = {
  type: 500
};

export default Error;
