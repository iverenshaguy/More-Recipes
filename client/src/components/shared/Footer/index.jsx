import React from 'react';
import './Footer.scss';

/**
 * @exports
 * @function Footer
 * @extends Component
 * @returns {JSX} Footer
 */
const Footer = () => (
  <footer className="footer">
    <div className="container">
      <p>Â© {(new Date()).getFullYear()} More Recipes</p>
    </div>
  </footer>
);

export default Footer;
