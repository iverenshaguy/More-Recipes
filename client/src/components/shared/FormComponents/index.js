import RenderInput from './RenderInput';
import RenderSelect from './RenderSelect';
import RenderFileInput from './RenderFileInput';
import RenderCheckBox from './RenderCheckBox';
import AddRecipeCompA from './AddRecipeCompA';
import AddRecipeCompB from './AddRecipeCompB';

export default {
  RenderInput,
  RenderSelect,
  RenderCheckBox,
  AddRecipeCompA,
  AddRecipeCompB,
  RenderFileInput
};
