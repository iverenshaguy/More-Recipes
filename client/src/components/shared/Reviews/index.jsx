import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import ReactStars from 'react-stars';
import Pagination from '../Pagination';
import { multiReviewPropTypes } from '../../../helpers/proptypes';
import { fetchReviews } from '../../../actions/recipeReviews';
import MiniPreLoader from '../../shared/PreLoader/MiniPreLoader';
import getNoResultText from '../../../helpers/getNoResultText';
import AddReview from './AddReview';

/**
 * Returns Reviews Wrapper
 * @exports
 * @extends Component
 * @class Reviews
 * @param {array} reviews
 * @returns {JSX} Reviews
 */
class Reviews extends Component {
  static propTypes = {
    error: PropTypes.string,
    recipeId: PropTypes.number.isRequired,
    dispatch: PropTypes.func.isRequired,
    reviewing: PropTypes.bool.isRequired,
    ...multiReviewPropTypes
  }

  static defaultProps = {
    error: null
  }

  /**
   * @constructor
   * @memberof Reviews
   * @returns {nothing} - Returns nothing
   */
  constructor() {
    super();

    this.state = {
      currentPage: 1,
      limit: 10
    };

    this.getReviews = this.getReviews.bind(this);
    this.handlePageChange = this.handlePageChange.bind(this);
  }

  /**
   * @memberof Reviews
   * @returns {noting} Returns nothing
   */
  componentWillMount() {
    this.getReviews();
  }

  /**
   * @memberof Reviews
   * @returns {noting} Returns nothing
   */
  getReviews() {
    const { currentPage, limit } = this.state;
    this.props.dispatch(fetchReviews(this.props.recipeId, currentPage, limit));
  }

  /**
   * @memberof Reviews
   * @param {number} page
   * @returns {state} home
   */
  handlePageChange(page) {
    this.setState({
      currentPage: page
    }, () => this.getReviews());
  }

  /**
   * @memberof Reviews
   * @returns {JSX} Reviews
   */
  render() {
    const {
      reviews, reviewing, error, metadata
    } = this.props;

    return (
      <div className="py-5 p-xs-1 p-md-5" id="reviews">
        <div className="col-12">
          <h4>Reviews</h4>
          <hr />
        </div>
        {reviewing && <div className="text-center"><MiniPreLoader /></div>}
        {!reviewing && !!reviews && reviews.length === 0 && <div className="text-center no-reviews">{getNoResultText('REVIEWS RESULT')}</div>}
        {error && <div className="text-center error">Something happened, please retry.</div>}
        {!!reviews && reviews.length !== 0 && reviews.sort((a, b) => b.id - a.id).map(review => (
          <div className="px-5 pb-3 review-heading" key={review.id}>
            <div className="row">
              <div className="col-4 rounded-circle text-left">
                <img src={review.User.profilePic} alt="user" className="review-picture" />
              </div>
              <div className="col-8 mr-0 text-right">
                <p className="mb-0">{review.User.username}</p>
              </div>
            </div>
            <hr style={{ marginTop: '0.5em', marginBottom: '0.5em' }} />
            <ReactStars value={+review.rating} size={30} color2="#f1c40f" half edit={false} />
            <p className="text-justify">{review.comment}</p>
          </div>))}
        {reviews.length !== 0 && <Pagination
          items={reviews}
          onPageChange={this.handlePageChange}
          metadata={metadata}
        />}
      </div>
    );
  }
}

const mapStateToProps = state => ({
  reviews: state.singleRecipe.recipeReviews.reviews,
  metadata: state.singleRecipe.recipeReviews.metadata,
  error: state.singleRecipe.recipeReviews.error,
  reviewing: state.singleRecipe.recipeReviews.reviewing
});

export { AddReview, Reviews as ReviewsComponent };
export default connect(mapStateToProps)(Reviews);
