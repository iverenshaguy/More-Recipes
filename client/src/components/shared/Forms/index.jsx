import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Form as BootstrapForm, Button } from 'reactstrap';
import { asyncValidate, syncValidate } from '../../../helpers/validations';
import AuthForm from './AuthForm';
import { NormalAlert } from '../Alert';
import { arrayToObject } from '../../../utils';
import { formHelpers } from '../../../helpers';

/**
 * @exports
 * @class Form
 * @extends Component
 * @returns {component} Form
 */
class Form extends Component {
  static propTypes = {
    dispatch: PropTypes.func.isRequired,
    submitting: PropTypes.bool.isRequired,
    submitError: PropTypes.string,
    type: PropTypes.string.isRequired,
    meta: PropTypes.shape({
      title: PropTypes.string,
      btnText: PropTypes.string,
      extra: PropTypes.element
    }).isRequired
  };

  static defaultProps = {
    submitError: null
  }

  /**
   * @constructor
   * @memberof Form
   * @param {object} props - props
   * @returns {nothing} Return nothing
   */
  constructor(props) {
    super(props);
    const { type } = props;
    const { formFields } = formHelpers;
    const fields = formFields[type];

    this.state = {
      type,
      values: arrayToObject(fields, ''),
      touched: arrayToObject(fields, false),
      error: arrayToObject(fields, null),
      pristine: true,
      formValid: false,
      asyncValidating: false
    };

    this.handleBlur = this.handleBlur.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleFocus = this.handleFocus.bind(this);
    this.validateField = this.validateField.bind(this);
    this.validateForm = this.validateForm.bind(this);
  }

  /**
   * @memberof Form
   * @returns {nothing} Returns nothing
   */
  componentWillMount() {
    this.clearFormErrors();
  }

  /**
   * @memberof Form
   * @returns {nothing} Returns nothing
   */
  clearFormErrors() {
    const { type } = this.state;
    const { clearFormError } = formHelpers;

    this.props.dispatch(clearFormError[type]);
  }


  /**
   * @memberof Form
   * @param {object} event
   * @returns {nothing} Returns nothing
   */
  handleChange(event) {
    const { target } = event;
    const { value, name } = target;

    this.setState({
      values: { ...this.state.values, [name]: value },
      touched: { ...this.state.touched, [name]: true },
      pristine: false,
    }, () => { this.validateField(name, value); });
  }

  /**
   * @memberof Form
   * @returns {nothing} Returns nothing
   */
  handleFocus() {
    this.clearFormErrors();
  }

  /**
   * @memberof Form
   * @param {object} event
   * @returns {nothing} Returns nothing
   */
  handleBlur(event) {
    const { target } = event;
    const { value, name } = target;
    const { type } = this.state;

    this.setState({
      touched: { ...this.state.touched, [name]: true },
    }, () => { this.validateField(name); });

    setTimeout(() => {
      if ((name === 'email' || name === 'username') && this.state.error[name] === null) {
        // only set asyncvalidating when sync validation is completed
        this.setState({ asyncValidating: true }, () => {
          asyncValidate(type)(name, value)
            .then(() => { this.setState({ asyncValidating: false }); })
            .catch((error) => {
              this.setState({
                error: { ...this.state.error, [name]: error[name] },
                asyncValidating: false
              });
            });
        });
      }
    }, 500);
  }

  /**
   * @memberof Form
   * @param {string} field
   * @param {string} values
   * @returns {nothing} Returns nothing
   */
  validateField(field) {
    const { type } = this.state;
    const error = syncValidate(type)(field, this.state.values);

    if (error) {
      this.setState({
        error: {
          ...this.state.error,
          [field]: error
        }
      }, this.validateForm);
    } else {
      this.setState({
        error: {
          ...this.state.error,
          [field]: null
        }
      }, this.validateForm);
    }
  }

  /**
   * @memberof Form
   * @returns {nothing} Returns nothing
   */
  validateForm() {
    const formErrorArrayLength =
      Object.values(this.state.error).filter(value => value !== null).length;
    const formTouchedArrayLength =
      Object.values(this.state.touched).filter(value => value === true).length;
    const formFieldCount = Object.keys(this.state.values).length;

    if (formErrorArrayLength) {
      this.setState({
        formValid: false
      });
    } else if (!formErrorArrayLength && formTouchedArrayLength === formFieldCount) {
      this.setState({
        formValid: true
      });
    }
  }

  /**
   * @memberof Form
   * @param {object} event
   * @returns {nothing} Returns nothing
   */
  handleSubmit(event) {
    event.preventDefault();
    const { type } = this.state;
    const { formSubmitMapper } = formHelpers;

    this.props.dispatch(formSubmitMapper[type](this.state.values));
  }

  /**
   * @memberof Form
   * @returns {component} Form
   */
  render() {
    const {
      values, touched, error, pristine, formValid, asyncValidating
    } = this.state;

    const {
      submitting, submitError, type, meta
    } = this.props;

    const { title, btnText, extra } = meta;

    const formState = {
      values,
      touched,
      error,
      pristine,
      formValid,
      asyncValidating
    };

    const handlers = {
      handleChange: this.handleChange,
      handleBlur: this.handleBlur,
      handleFocus: this.handleFocus,
      handleSubmit: this.handleSubmit
    };

    return (
      <div>
        <h4 className="text-center">{title}</h4>
        <hr />
        {type !== 'login' &&
          <p className="text-muted mx-auto text-center">
            Fields marked
            <span className="text-danger">*</span> are important
          </p>}
        <BootstrapForm className="mt-4 mb-3 px-5" onSubmit={handlers.handleSubmit}>
          {submitError && (
            <NormalAlert color="danger">
              <p className="text-center mb-0">{submitError}</p>
            </NormalAlert>
          )}
          {(type === 'signup' || type === 'login') && <AuthForm type={type} state={formState} handlers={handlers} />}
          <Button className="btn-block mt-0" disabled={!formValid || pristine || submitting}>
            {btnText}
          </Button>
        </BootstrapForm>
        {extra && extra}
      </div>);
  }
}


export { Form as FormComponent };

export default connect()(Form);
