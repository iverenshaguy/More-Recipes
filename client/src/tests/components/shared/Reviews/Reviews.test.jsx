import React from 'react';
import { ReviewsComponent as Reviews } from '../../../../components/shared/Reviews';

const dispatchMock = jest.fn();

const setup = () => {
  const props = {
    recipeId: 5,
    dispatch: dispatchMock,
    reviews: [
      {
        id: 56,
        rating: 5,
        comment: 'Very Good',
        recipeId: 5,
        User: {
          id: 2,
          username: 'user',
          profilePic: null
        }
      },
      {
        id: 3,
        rating: 4,
        comment: 'Great',
        recipeId: 5,
        User: {
          id: 4,
          username: 'user',
          profilePic: null
        }
      },
      {
        id: 78,
        rating: 5,
        comment: 'Excellent',
        recipeId: 5,
        User: {
          id: 6,
          username: 'user6',
          profilePic: null
        }
      }
    ],
    metadata: {
      firstPage: 1,
      lastPage: 2,
      page: 1,
      itemsPerPage: 2,
      pages: [1, 2],
      totalCount: 3,
    },
    error: null,
    reviewing: false,
    user: {
      id: 43,
      firstName: 'Ivy',
      lastName: 'Shaguy',
      username: 'ivy'
    }
  };

  const shallowWrapper = shallow(<Reviews {...props} />);

  const mountedWrapper = mount(<Reviews {...props} />);

  return { props, mountedWrapper, shallowWrapper };
};

describe('Reviews', () => {
  it('renders correctly', () => {
    const { shallowWrapper } = setup();
    expect(toJson(shallowWrapper)).toMatchSnapshot();
  });

  it('calls dispatch when mounted', () => {
    const { mountedWrapper } = setup();
    expect(dispatchMock).toHaveBeenCalled();
    mountedWrapper.unmount();
  });

  it('renders MiniPreloader when loading', () => {
    const { props } = setup();
    const shallowWrapper = shallow(<Reviews {...props} reviewing />);

    expect(shallowWrapper.find('MiniPreLoader').length).toBeTruthy();
  });

  it('renders message when there are no reviews', () => {
    const { props } = setup();
    const shallowWrapper = shallow(<Reviews {...props} reviews={[]} />);
    expect(shallowWrapper.find('.no-reviews').length).toBeTruthy();
    expect(shallowWrapper.find('.no-reviews').text()).toEqual('There are no reviews for this recipe');
  });

  it('renders error message when there is an error', () => {
    const { props } = setup();
    const shallowWrapper = shallow(<Reviews {...props} error="This is an error" />);

    expect(shallowWrapper.find('.error').length).toBeTruthy();
    expect(shallowWrapper.find('.error').text()).toEqual('Something happened, please retry.');
  });

  it('renders pagination when reviews are more than current page', () => {
    const state = {
      currentPage: 1,
      limit: 2
    };

    const { mountedWrapper } = setup();
    const handlePageChangeSpy = jest.spyOn(mountedWrapper.instance(), 'handlePageChange');

    mountedWrapper.setState(state);
    mountedWrapper.find('a.page-link').at(3).simulate('click');

    expect(handlePageChangeSpy).toHaveBeenCalled();
    expect(mountedWrapper.instance().state.currentPage).toEqual(2);
    mountedWrapper.unmount();
  });
});
