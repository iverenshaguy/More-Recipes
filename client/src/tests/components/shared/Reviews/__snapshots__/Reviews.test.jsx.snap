// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reviews renders correctly 1`] = `
<div
  className="py-5 p-xs-1 p-md-5"
  id="reviews"
>
  <div
    className="col-12"
  >
    <h4>
      Reviews
    </h4>
    <hr />
  </div>
  <div
    className="px-5 pb-3 review-heading"
    key="78"
  >
    <div
      className="row"
    >
      <div
        className="col-4 rounded-circle text-left"
      >
        <img
          alt="user"
          className="review-picture"
          src={null}
        />
      </div>
      <div
        className="col-8 mr-0 text-right"
      >
        <p
          className="mb-0"
        >
          user6
        </p>
      </div>
    </div>
    <hr
      style={
        Object {
          "marginBottom": "0.5em",
          "marginTop": "0.5em",
        }
      }
    />
    <ReactStars
      char="★"
      color1="gray"
      color2="#f1c40f"
      count={5}
      edit={false}
      half={true}
      onChange={[Function]}
      size={30}
      value={5}
    />
    <p
      className="text-justify"
    >
      Excellent
    </p>
  </div>
  <div
    className="px-5 pb-3 review-heading"
    key="56"
  >
    <div
      className="row"
    >
      <div
        className="col-4 rounded-circle text-left"
      >
        <img
          alt="user"
          className="review-picture"
          src={null}
        />
      </div>
      <div
        className="col-8 mr-0 text-right"
      >
        <p
          className="mb-0"
        >
          user
        </p>
      </div>
    </div>
    <hr
      style={
        Object {
          "marginBottom": "0.5em",
          "marginTop": "0.5em",
        }
      }
    />
    <ReactStars
      char="★"
      color1="gray"
      color2="#f1c40f"
      count={5}
      edit={false}
      half={true}
      onChange={[Function]}
      size={30}
      value={5}
    />
    <p
      className="text-justify"
    >
      Very Good
    </p>
  </div>
  <div
    className="px-5 pb-3 review-heading"
    key="3"
  >
    <div
      className="row"
    >
      <div
        className="col-4 rounded-circle text-left"
      >
        <img
          alt="user"
          className="review-picture"
          src={null}
        />
      </div>
      <div
        className="col-8 mr-0 text-right"
      >
        <p
          className="mb-0"
        >
          user
        </p>
      </div>
    </div>
    <hr
      style={
        Object {
          "marginBottom": "0.5em",
          "marginTop": "0.5em",
        }
      }
    />
    <ReactStars
      char="★"
      color1="gray"
      color2="#f1c40f"
      count={5}
      edit={false}
      half={true}
      onChange={[Function]}
      size={30}
      value={4}
    />
    <p
      className="text-justify"
    >
      Great
    </p>
  </div>
  <Pagination
    items={
      Array [
        Object {
          "User": Object {
            "id": 6,
            "profilePic": null,
            "username": "user6",
          },
          "comment": "Excellent",
          "id": 78,
          "rating": 5,
          "recipeId": 5,
        },
        Object {
          "User": Object {
            "id": 2,
            "profilePic": null,
            "username": "user",
          },
          "comment": "Very Good",
          "id": 56,
          "rating": 5,
          "recipeId": 5,
        },
        Object {
          "User": Object {
            "id": 4,
            "profilePic": null,
            "username": "user",
          },
          "comment": "Great",
          "id": 3,
          "rating": 4,
          "recipeId": 5,
        },
      ]
    }
    metadata={
      Object {
        "firstPage": 1,
        "itemsPerPage": 2,
        "lastPage": 2,
        "page": 1,
        "pages": Array [
          1,
          2,
        ],
        "totalCount": 3,
      }
    }
    onPageChange={[Function]}
  />
</div>
`;
